doc/man/doveadm.1.in
.\" Copyright (c) 2010-2018 Dovecot authors, see the included COPYING file
.TH DOVEADM 1 "2014-10-07" "Dovecot v2.3" "Dovecot"
.SH NAME
doveadm \- Dovecot\(aqs administration utility
.\"------------------------------------------------------------------------
.SH SYNOPSIS
.BR doveadm " [" \-Dv "] [" \-f
.IR formatter ]
.IR command " [" command_options "] [" command_arguments ]
.\"------------------------------------------------------------------------
.SH DESCRIPTION
.B doveadm
is the Dovecot administration tool. It can be used to manage various parts
of Dovecot, as well as access users\(aq mailboxes.
.br
Execute
.BR doveadm\ help ,
for a command usage listing.
.\"------------------------------------------------------------------------
@INCLUDE:global-options-formatter@
.\"------------------------------------------------------------------------
.SH COMMANDS
.\"------------------------------------------------------------------------
These commands work directly with Dovecot\(aqs master process.
.\"-------------------------------------
.SS reload
.B doveadm reload
.PP
Force
.BR dovecot (1)
to reload the configuration.
.\"------------------------------------------------------------------------
.SS stop
.B doveadm stop
.PP
Stop
.BR dovecot (1)
and all its child processes.
.\"------------------------------------------------------------------------
.SS ADDITIONAL MASTER COMMANDS
.TP
.B doveadm config
.BR doveadm\-config (1),
Dump Dovecot configuration.
.\"-------------------------------------
.TP
.B doveadm director
.BR doveadm\-director (1),
Manage Dovecot directors (if used by proxy servers).
.\"-------------------------------------
.TP
.B doveadm exec
.BR doveadm\-exec (1),
Easily execute commands from Dovecot\(aqs libexec_dir.
.\"-------------------------------------
.TP
.B doveadm instance
.BR doveadm\-instance (1),
Manage the list of running Dovecot instances.
.\"-------------------------------------
.TP
.B doveadm kick
.BR doveadm\-kick (1),
Disconnect users by user name and/or IP address.
.\"-------------------------------------
.TP
.B doveadm log
.BR doveadm\-log (1),
Locate, test or reopen Dovecot\(aqs log files.
.\"-------------------------------------
.TP
.B doveadm penalty
.BR doveadm\-penalty (1),
Show current penalties.
.\"-------------------------------------
.TP
.B doveadm proxy
.BR doveadm\-proxy (1),
Handle Dovecot proxy connections.
.\"-------------------------------------
.TP
.B doveadm who
.BR doveadm\-who (1);
Show who is logged in to the Dovecot server.
.\"------------------------------------------------------------------------
.SS AUTHENTICATION COMMANDS
.\"-------------------------------------
.TP
.B doveadm auth
.BR doveadm\-auth (1),
Test authentication for a user.
.\"-------------------------------------
.TP
.B doveadm pw
.BR doveadm\-pw (1),
Dovecot\(aqs password hash generator.
.\"-------------------------------------
.TP
.B doveadm user
.BR doveadm\-user (1),
Perform a user lookup in Dovecot\(aqs userdbs
.\"------------------------------------------------------------------------
.SS MAILBOX COMMANDS
.TP
.B doveadm acl
.BR doveadm\-acl (1),
Manage Access Control List (ACL).
.\"-------------------------------------
.TP
.B doveadm altmove
.BR doveadm\-altmove (1),
Move matching mails to the alternative storage.
.\"-------------------------------------
.TP
.B doveadm backup
.BR doveadm\-backup (1),
Dovecot\(aqs one\-way mailbox synchronization utility.
.\"-------------------------------------
.TP
.B doveadm copy
.BR doveadm\-copy (1),
Copy messages matching the given search query into another mailbox.
.\"-------------------------------------
.TP
.B doveadm deduplicate
.BR doveadm\-deduplicate (1),
Expunge duplicate messages.
.\"-------------------------------------
.TP
.B doveadm dump
.BR doveadm\-dump (1),
Dump the content of Dovecot\(aqs binary mailbox index/log.
.\"-------------------------------------
.TP
.B doveadm expunge
.BR doveadm\-expunge (1),
Expunge messages matching given search query.
.\"-------------------------------------
.TP
.B doveadm fetch
.BR doveadm\-fetch (1),
Fetch messages matching given search query.
.\"-------------------------------------
.TP
.B doveadm flags
.BR doveadm\-flags (1),
Add, remove or replace messages\(aq flags.
.\"-------------------------------------
.TP
.B doveadm force\-resync
.BR doveadm\-force\-resync (1),
Repair broken mailboxes, in case Dovecot doesn\(aqt automatically do that.
.\"-------------------------------------
.TP
.B doveadm fs
.BR doveadm\-fs (1),
Interact with the abstract mail storage filesystem.
.\"-------------------------------------
.TP
.B doveadm fts
.BR doveadm\-fts (1),
Manipulate the Full Text Search (FTS) index.
.\"-------------------------------------
.TP
.B doveadm import
.BR doveadm\-import (1),
Import messages matching given search query.
.\"-------------------------------------
.TP
.B doveadm index
.BR doveadm\-index (1),
Index messages in a given mailbox.
.\"-------------------------------------
.TP
.B doveadm mailbox
.BR doveadm\-mailbox (1),
Various commands related to handling mailboxes.
.\"-------------------------------------
.TP
.B doveadm mailbox cryptokey
.BR doveadm\-mailbox\-cryptokey (1),
Mail crypt plugin management.
.\"-------------------------------------
.TP
.B doveadm move
.BR doveadm\-move (1)
Move messages matching the given search query into another mailbox.
.\"-------------------------------------
.TP
.B doveadm purge
.BR doveadm\-purge (1)
Remove messages with refcount=0 from mdbox files.
.\"-------------------------------------
.TP
.B doveadm quota
.BR doveadm\-quota (1),
Initialize/recalculate or show current quota usage.
.\"-------------------------------------
.TP
.B doveadm rebuild
.BR doveadm\-rebuild (1),
Rebuild index metadata from message data.
.\"-------------------------------------
.TP
.B doveadm replicator
.BR doveadm\-replicator (1),
Manage users\(aq mail replication.
.\"-------------------------------------
.TP
.B doveadm save
.BR doveadm\-save (1),
Save email to users\(aq mailboxes.
.\"-------------------------------------
.TP
.B doveadm search
.BR doveadm\-search (1),
Show a list of mailbox GUIDs and message UIDs matching given search query.
.\"-------------------------------------
.TP
.B doveadm stats
.BR doveadm\-stats (1),
Inspect or reset stats.
.\"-------------------------------------
.TP
.B doveadm sync
.BR doveadm\-sync (1),
Dovecot\(aqs two\-way mailbox synchronization utility.
.\"------------------------------------------------------------------------
.SH "EXIT STATUS"
.B doveadm
will exit with one of the following values:
.TP 4
.B 0
Selected command was executed successful.
.TP
.B >0
Command failed in some way.
.\"------------------------------------------------------------------------
.SH ENVIRONMENT
.TP
.B USER
This environment variable is used to determine the
.I user
if a command accepts a
.I user
but none was specified.
.\"------------------------------------------------------------------------
.SH FILES
.TP
.I @pkgsysconfdir@/dovecot.conf
Dovecot\(aqs main configuration file.
.TP
.I @pkgsysconfdir@/conf.d/10\-mail.conf
Mailbox locations and namespaces.
.TP
.I @pkgsysconfdir@/conf.d/90\-plugin.conf
Plugin specific settings.
.\"------------------------------------------------------------------------
@INCLUDE:reporting-bugs@
.\"------------------------------------------------------------------------
.SH SEE ALSO
.BR doveadm\-help (1),
.BR doveconf (1),
.BR dovecot (1),
.BR doveadm\-search\-query (7)


NEWS
v2.3.19 2022-05-10  Aki Tuomi <aki.tuomi@open-xchange.com>

	+ Added mail_user_session_finished event, which is emitted when the mail
	  user session is finished (e.g. imap, pop3, lmtp). It also includes
	  fields with some process statistics information.
	  See https://doc.dovecot.org/admin_manual/list_of_events/ for more
	  information.
	+ Added process_shutdown_filter setting. When an event matches the filter,
	  the process will be shutdown after the current connection(s) have
	  finished. This is intended to reduce memory usage of long-running imap
	  processes that keep a lot of memory allocated instead of freeing it to
	  the OS.
	+ auth: Add cache hit indicator to auth passdb/userdb finished events.
	  See https://doc.dovecot.org/admin_manual/list_of_events/ for more
	  information.
	+ doveadm deduplicate: Performance is improved significantly.
	+ imapc: COPY commands were sent one mail at a time to the remote IMAP
	  server. Now the copying is buffered, so multiple mails can be copied
	  with a single COPY command.
	+ lib-lua: Add a Lua interface to Dovecot's HTTP client library. See
	  https://doc.dovecot.org/admin_manual/lua/ for more information.
	- auth: Cache lookup would use incorrect cache key after username change.
	- auth: Improve handling unexpected LDAP connection errors/hangs.
	  Try to fix up these cases by reconnecting to the LDAP server and
	  aborting LDAP requests earlier.
	- auth: Process crashed if userdb iteration was attempted while auth-workers
	  were already full handling auth requests.
	- auth: db-oauth2: Using %{oauth2:name} variables caused unnecessary
	  introspection requests.
	- dict: Timeouts may have been leaked at deinit.
	- director: Ring may have become unstable if a backend's tag was changed.
	  It could also have caused director process to crash.
	- doveadm kick: Numeric parameter was treated as IP address.
	- doveadm: Proxying can panic when flushing print output. Fixes
	  Panic: file ioloop.c: line 865 (io_loop_destroy): assertion failed:
	  (ioloop == current_ioloop).
	- doveadm sync: BROKENCHAR was wrongly changed to '_' character when
	  migrating mailboxes. This was set by default to %, so any mailbox
	  names containing % characters were modified to "_25".
	- imapc: Copying or moving mails with doveadm to an imapc mailbox could
	  have produced "Error: Syncing mailbox '[...]' failed" Errors. The
	  operation itself succeeded but attempting to sync the destination
	  mailbox failed.
	- imapc: Prevent index log synchronization errors when two or more imapc
	  sessions are adding messages to the same mailbox index files, i.e.
	  INDEX=MEMORY is not used.
	- indexer: Process was slowly leaking memory for each indexing request.
	- lib-fts: fts header filters caused binary content to be sent to the
	  indexer with non-default configuration.
	- doveadm-server: Process could hang in some situations when printing
	  output to TCP client, e.g. when printing doveadm sync state.
	- lib-index: dovecot.index.log files were often read and parsed entirely,
	  rather than only the parts that were actually necessary. This mainly
	  increased CPU usage.
	- lmtp-proxy: Session ID forwarding would cause same session IDs being
	  used when delivering same mail to multiple backends.
	- log: Log prefix update may have been lost if log process was busy.
	  This could have caused log prefixes to be empty or in some cases
	  reused between sessions, i.e. log lines could have been logged for the
	  wrong user/session.
	- mail_crypt: Plugin crashes if it's loaded only for some users. Fixes
	  Panic: Module context mail_crypt_user_module missing.
	- mail_crypt: When LMTP was delivering mails to both recipients with mail
	  encryption enabled and not enabled, the non-encrypted recipients may
	  have gotten mails encrypted anyway. This happened when the first
	  recipient was encrypted (mail_crypt_save_version=2) and the 2nd
	  recipient was not encrypted (mail_crypt_save_version=0).
	- pop3: Session would crash if empty line was sent.
	- stats: HTTP server leaked memory.
	- submission-login: Long credentials, such as OAUTH2 tokens, were refused
	  during SASL interactive due to submission server applying line length
	  limits.
	- submission-login: When proxying to remote host, authentication was not
	  using interactive SASL when logging in using long credentials such as
	  OAUTH2 tokens. This caused authentication to fail due to line length
	  constraints in SMTP protocol.
	- submission: Terminating the client connection with QUIT command after
	  mail transaction is started with MAIL command and before it is
	  finished with DATA/BDAT can cause a segfault crash.
	- virtual: doveadm search queries with mailbox-guid as the only parameter
	  crashes: Panic: file virtual-search.c: line 77 (virtual_search_get_records):
	  assertion failed: (result != 0)

v2.3.18 2022-02-03  Aki Tuomi <aki.tuomi@open-xchange.com>

	* Removed mail_cache_lookup_finished event. This event wasn't especially
	  useful, but it increased CPU usage significantly.
	* fts: Don't index inline base64 encoded content in FTS indexes using
	  the generic tokenizer. This reduces the FTS index sizes by removing
	  input that is very unlikely to be searched for. See
	  https://doc.dovecot.org/configuration_manual/fts/tokenization for
	  details on how base64 is detected. Only applies when using libfts.
	* lmtp: Session IDs are now preserved through proxied connections, so
	  LMTP sessions can be tracked. This slightly changes the LMTP session
	  ID format by appending ":Tn" (transaction), ":Pn" (proxy connection)
	  and ":Rn" (recipient) counters after the session ID prefix.
	+ Events now have "reason_code" field, which can provide a list of
	  reasons why the event is happening. See
	  https://doc.dovecot.org/admin_manual/event_reasons/
	+ New events are added. See https://doc.dovecot.org/admin_manual/list_of_events/
	+ fts: Added fts_header_excludes and fts_header_includes settings to

src/doveadm/doveadm.c
/* Copyright (c) 2009-2018 Dovecot authors, see the included COPYING file */

#include "lib.h"
#include "array.h"
#include "str.h"
#include "sort.h"
#include "ostream.h"
#include "env-util.h"
#include "execv-const.h"
#include "dict.h"
#include "master-service-private.h"
#include "master-service-settings.h"
#include "settings-parser.h"
#include "doveadm-print-private.h"
#include "doveadm-dump.h"
#include "doveadm-mail.h"
#include "doveadm-settings.h"
#include "doveadm-dsync.h"
#include "doveadm.h"

#include <unistd.h>

const struct doveadm_print_vfuncs *doveadm_print_vfuncs_all[] = {
	&doveadm_print_flow_vfuncs,
	&doveadm_print_tab_vfuncs,
	&doveadm_print_table_vfuncs,
	&doveadm_print_pager_vfuncs,
	&doveadm_print_json_vfuncs,
	&doveadm_print_formatted_vfuncs,
	NULL
};

int doveadm_exit_code = 0;

static void failure_exit_callback(int *status)
{
	enum fatal_exit_status fatal_status = *status;

	switch (fatal_status) {
	case FATAL_LOGWRITE:
	case FATAL_LOGERROR:
	case FATAL_LOGOPEN:
	case FATAL_OUTOFMEM:
	case FATAL_EXEC:
	case FATAL_DEFAULT:
		*status = EX_TEMPFAIL;
		break;
	}
}

static void
doveadm_usage_compress_lines(FILE *out, const char *str, const char *prefix)
{
	const char *cmd, *args, *p, *short_name, *sub_name;
	const char *prev_name = "", *prev_sub_name = "";
	const char **lines;
	unsigned int i, count;
	size_t prefix_len = strlen(prefix);

	/* split lines */
	lines = (void *)p_strsplit(pool_datastack_create(), str, "\n");
	for (count = 0; lines[count] != NULL; count++) ;

	/* sort lines */
	i_qsort(lines, count, sizeof(*lines), i_strcmp_p);

	/* print lines, compress subcommands into a single line */
	for (i = 0; i < count; i++) {
		args = strchr(lines[i], '\t');
		if (args == NULL) {
			cmd = lines[i];
			args = "";
		} else {
			cmd = t_strdup_until(lines[i], args);
			args++;
		}
		if (*prefix != '\0') {
			if (strncmp(cmd, prefix, prefix_len) != 0 ||
			    cmd[prefix_len] != ' ')
				continue;
			cmd += prefix_len + 1;
		}

		p = strchr(cmd, ' ');
		if (p == NULL) {
			if (*prev_name != '\0') {
				fprintf(out, "\n");
				prev_name = "";
			}
			fprintf(out, USAGE_CMDNAME_FMT" %s\n", cmd, args);
		} else {
			short_name = t_strdup_until(cmd, p);
			if (strcmp(prev_name, short_name) != 0) {
				if (*prev_name != '\0')
					fprintf(out, "\n");
				fprintf(out, USAGE_CMDNAME_FMT" %s",
					short_name, t_strcut(p + 1, ' '));
				prev_name = short_name;
				prev_sub_name = "";
			} else {
				sub_name = t_strcut(p + 1, ' ');
				if (strcmp(prev_sub_name, sub_name) != 0) {
					fprintf(out, "|%s", sub_name);
					prev_sub_name = sub_name;
				}
			}
		}
	}
	if (*prev_name != '\0')
		fprintf(out, "\n");
}

static void ATTR_NORETURN
usage_prefix(const char *prefix)
{
	const struct doveadm_cmd_ver2 *cmd2;
	string_t *str = t_str_new(1024);

	fprintf(stderr, "usage: doveadm [-Dv] [-f <formatter>] ");
	if (*prefix != '\0')
		fprintf(stderr, "%s ", prefix);
	fprintf(stderr, "<command> [<args>]\n");

	array_foreach(&doveadm_cmds_ver2, cmd2) {
		if ((cmd2->flags & CMD_FLAG_HIDDEN) == 0)
			str_printfa(str, "%s\t%s\n", cmd2->name, cmd2->usage);
	}

	doveadm_usage_compress_lines(stderr, str_c(str), prefix);

	lib_exit(EX_USAGE);
}

void usage(void)
{
	usage_prefix("");
}

void help_ver2(const struct doveadm_cmd_ver2 *cmd)
{
	fprintf(stderr, "doveadm %s %s\n", cmd->name, cmd->usage);
	lib_exit(EX_USAGE);
}

static void cmd_help(struct doveadm_cmd_context *cctx)
{
	const char *cmd, *man_argv[3];

	if (!doveadm_cmd_param_str(cctx, "cmd", &cmd))
		usage_prefix("");

	env_put("MANPATH", MANDIR);
	man_argv[0] = "man";
	man_argv[1] = t_strconcat("doveadm-", cmd, NULL);
	man_argv[2] = NULL;
	execvp_const(man_argv[0], man_argv);
}

static struct doveadm_cmd_ver2 doveadm_cmd_help = {
	.name = "help",
	.cmd = cmd_help,
	.usage = "[<cmd>]",
DOVEADM_CMD_PARAMS_START
DOVEADM_CMD_PARAM('\0', "cmd", CMD_PARAM_STR, CMD_PARAM_FLAG_POSITIONAL)
DOVEADM_CMD_PARAMS_END
};

static void cmd_config(struct doveadm_cmd_context *cctx)
{
	const char *const *args, **argv;

	if (!doveadm_cmd_param_array(cctx, "args", &args))
		args = NULL;

	env_put(MASTER_CONFIG_FILE_ENV,
		master_service_get_config_path(master_service));

	unsigned int len = str_array_length(args);
	argv = t_new(const char *, len + 2);
	argv[0] = BINDIR"/doveconf";
	if (len > 0) {
		i_assert(args != NULL);
		memcpy(argv+1, args, len * sizeof(args[0]));
	}
	execv_const(argv[0], argv);
}

static struct doveadm_cmd_ver2 doveadm_cmd_config = {
	.name = "config",
	.cmd = cmd_config,
	.usage = "[doveconf parameters]",
	.flags = CMD_FLAG_NO_OPTIONS,
DOVEADM_CMD_PARAMS_START
DOVEADM_CMD_PARAM('\0', "args", CMD_PARAM_ARRAY, CMD_PARAM_FLAG_POSITIONAL)
DOVEADM_CMD_PARAMS_END
};

static void cmd_exec(struct doveadm_cmd_context *cctx);
static struct doveadm_cmd_ver2 doveadm_cmd_exec = {
	.name = "exec",
	.cmd = cmd_exec,
	.usage = "<binary> [binary parameters]",
	.flags = CMD_FLAG_NO_OPTIONS,
DOVEADM_CMD_PARAMS_START
DOVEADM_CMD_PARAM('\0', "binary", CMD_PARAM_STR, CMD_PARAM_FLAG_POSITIONAL)
DOVEADM_CMD_PARAM('\0', "args", CMD_PARAM_ARRAY, CMD_PARAM_FLAG_POSITIONAL)
DOVEADM_CMD_PARAMS_END
};

static void cmd_exec(struct doveadm_cmd_context *cctx)
{
	const char *path, *binary, *const *args, **argv;

	if (!doveadm_cmd_param_str(cctx, "binary", &binary))
		help_ver2(&doveadm_cmd_exec);
	if (!doveadm_cmd_param_array(cctx, "args", &args))
		args = NULL;

	path = t_strdup_printf("%s/%s", doveadm_settings->libexec_dir, binary);

	unsigned int len = str_array_length(args);
	argv = t_new(const char *, len + 2);
	argv[0] = path;
	if (len > 0) {
		i_assert(args != NULL);
		memcpy(argv+1, args, len * sizeof(args[0]));
	}
	execv_const(argv[0], argv);
}

static bool doveadm_has_subcommands(const char *cmd_name)
{
	const struct doveadm_cmd_ver2 *cmd2;
	size_t len = strlen(cmd_name);

	array_foreach(&doveadm_cmds_ver2, cmd2) {
		if (strncmp(cmd2->name, cmd_name, len) == 0 &&
		    cmd2->name[len] == ' ')
			return TRUE;
	}
	return FALSE;
}

static struct doveadm_cmd_ver2 *doveadm_cmdline_commands_ver2[] = {
	&doveadm_cmd_config,
	&doveadm_cmd_dump,
	&doveadm_cmd_exec,
	&doveadm_cmd_help,
	&doveadm_cmd_oldstats_top_ver2,
	&doveadm_cmd_pw,
	&doveadm_cmd_zlibconnect,
};

int main(int argc, char *argv[])
{
	enum master_service_flags service_flags =
		MASTER_SERVICE_FLAG_STANDALONE |
		MASTER_SERVICE_FLAG_KEEP_CONFIG_OPEN |
		MASTER_SERVICE_FLAG_NO_SSL_INIT |
		MASTER_SERVICE_FLAG_NO_INIT_DATASTACK_FRAME;
	struct doveadm_cmd_context cctx;
	const char *cmd_name;
	unsigned int i;
	bool quick_init = FALSE;
	int c;

	i_zero(&cctx);
	cctx.conn_type = DOVEADM_CONNECTION_TYPE_CLI;

	i_set_failure_exit_callback(failure_exit_callback);

	/* "+" is GNU extension to stop at the first non-option.
	   others just accept -+ option. */
	master_service = master_service_init("doveadm", service_flags,
					     &argc, &argv, "+Df:hv");
	while ((c = master_getopt(master_service)) > 0) {
		switch (c) {
		case 'D':
			doveadm_debug = TRUE;
			doveadm_verbose = TRUE;
			break;
		case 'f':
			doveadm_print_init(optarg);
			break;
		case 'h':
			doveadm_print_hide_titles = TRUE;
			break;
		case 'v':
			doveadm_verbose = TRUE;
			break;
		default:
			return FATAL_DEFAULT;
		}
	}
	cmd_name = argv[optind];

	if (cmd_name != NULL && strcmp(cmd_name, "help") == 0 &&
	    argv[optind+1] != NULL) {
		/* "help cmd" doesn't need any configuration */
		quick_init = TRUE;
	}
	master_service_init_log(master_service);

	doveadm_settings_init();
	doveadm_cmds_init();
	for (i = 0; i < N_ELEMENTS(doveadm_cmdline_commands_ver2); i++)
		doveadm_cmd_register_ver2(doveadm_cmdline_commands_ver2[i]);
	doveadm_register_auth_commands();

	if (cmd_name != NULL && (quick_init ||
				 strcmp(cmd_name, "config") == 0 ||
				 strcmp(cmd_name, "stop") == 0 ||
				 strcmp(cmd_name, "reload") == 0)) {
		/* special case commands: even if there is something wrong
		   with the config (e.g. mail_plugins), don't fail these
		   commands */
		if (strcmp(cmd_name, "help") != 0)
			doveadm_read_settings();
		quick_init = TRUE;
	} else {
		quick_init = FALSE;
		doveadm_print_ostream = o_stream_create_fd(STDOUT_FILENO, 0);
		o_stream_set_no_error_handling(doveadm_print_ostream, TRUE);
		o_stream_cork(doveadm_print_ostream);
		doveadm_dump_init();
		doveadm_mail_init();
		dict_drivers_register_builtin();
		doveadm_load_modules();

		/* read settings only after loading doveadm plugins, which
		   may modify what settings are read */
		doveadm_read_settings();
		if (doveadm_debug && getenv("LOG_STDERR_TIMESTAMP") == NULL)
			i_set_failure_timestamp_format(master_service->set->log_timestamp);
		master_service_init_stats_client(master_service, TRUE);
		/* Load mail_plugins */
		doveadm_mail_init_finish();
		/* kludgy: Load the rest of the doveadm plugins after
		   mail_plugins have been loaded. */
		doveadm_load_modules();

		if (cmd_name == NULL) {
			/* show usage after registering all plugins */
			usage_prefix("");
		}
	}

	argc -= optind;
	argv += optind;
	i_getopt_reset();

	master_service_init_finish(master_service);
	if (!doveadm_debug) {
		/* disable debugging unless -D is given */
		i_set_debug_file("/dev/null");
	}

	/* this has to be done here because proctitle hack can break
	   the env pointer */
	cctx.username = getenv("USER");

	if (!doveadm_cmdline_try_run(cmd_name, argc, (const char**)argv, &cctx)) {
		if (doveadm_has_subcommands(cmd_name))
			usage_prefix(cmd_name);
		if (doveadm_has_unloaded_plugin(cmd_name)) {
			i_fatal("Unknown command '%s', but plugin %s exists. "
				"Try to set mail_plugins=%s",
				cmd_name, cmd_name, cmd_name);
		}
		usage();
	}
	if (cctx.referral != NULL)
		i_fatal("Command requested referral: %s", cctx.referral);

	if (!quick_init) {
		doveadm_mail_deinit();
		doveadm_dump_deinit();
		doveadm_unload_modules();
		dict_drivers_unregister_builtin();
		doveadm_print_deinit();
		o_stream_unref(&doveadm_print_ostream);
	}
	doveadm_cmds_deinit();
	doveadm_settings_deinit();
	master_service_deinit(&master_service);
	return doveadm_exit_code;
}
